mmcv/runner/dist_utils.py

```diff
@@ -33,10 +33,13 @@ def _init_dist_pytorch(backend, **kwargs):
 
 
 def _init_dist_mpi(backend, **kwargs):
-    # TODO: use local_rank instead of rank % num_gpus
-    rank = int(os.environ['OMPI_COMM_WORLD_RANK'])
-    num_gpus = torch.cuda.device_count()
-    torch.cuda.set_device(rank % num_gpus)
+    local_rank = int(os.environ['OMPI_COMM_WORLD_LOCAL_RANK'])
+    torch.cuda.set_device(local_rank)
+    if 'MASTER_PORT' not in os.environ:
+        # 29500 is torch.distributed default port
+        os.environ['MASTER_PORT'] = '29500'
+    os.environ['WORLD_SIZE'] = os.environ['OMPI_COMM_WORLD_SIZE']
+    os.environ['RANK'] = os.environ['OMPI_COMM_WORLD_RANK']
```

Comment:
Could you explain why we need to set these environment variables here?

Reply:
I saw the `--launcher mpi` option of `tools/train.py` but couldn't find documentation about it. 

I tried launching a multi-node MMDetection training job using `mpirun` (arguments similar to those of Horovod) and `tools/train.py $CONFIG --launcher mpi`, with one process per GPU. An error occurred which told that environmental variables like `RANK` are missing. So I set their values using OpenMPI's counterparts, and the training job succeeded.

Reply:
Got it. Thanks for your description.

Reply:
Can mpi backend be launched by `torch.distributed.launch?`. If possible， would these environment variables be overwritten?

Reply:
I assumed mpi backend is to be launched by `mpirun` only, because the original code already uses `OMPI_COMM_WORLD_RANK` which is an environmental variable offered by MPI but not `torch.distributed.launch`.

Reply:
@bryandeng, agree with you.

Reply:
Using `mpirun`  with `nccl` backend should also set environment variables as `slurm` does, refer to https://pytorch.org/tutorials/intermediate/dist_tuto.html#initialization-methods 
However, such variables might not needed for `mpi` backend, because they are already assigned by MPI runtime.

Reply:
@bryandeng, I think it's better to add `MASTER_ADDR` environment variable according to the explanation from @yingfhu.

Reply:
@teamwong111 I understand what you and @yingfhu mean.

I also wrote a launching script which contains `mpirun` and will pass on `MASTER_ADDR` to the downstream script. But I think such a launching script should be put in maybe MMDetection's repository and be named like `tools/mpi_train.sh`, not in `mmcv`'s repository.

In this pull request, I'm submitting code changes related to `mmcv`. I would be very grateful if you could communicate with your whole team about such a `tools/mpi_train.sh` launching script, as it would be shared by various MM toolboxes just like `tools/slurm_train.sh`. Please tell me where I should contribute this script.

Reply:
`MASTER_ADDR` is also needed in here for completeness, but openmpi doesn't seem to have environment variable resembles to `SLURM_NODELIST` in `slurm`, should add extra param here for `MASTER_ADDR` or at least add a warning to inform users to add this environment variable by themselves.

Reply:
> @teamwong111 I understand what you and @yingfhu mean.
> 
> I also wrote a launching script which contains `mpirun` and will pass on `MASTER_ADDR` to the downstream script. But I think such a launching script should be put in maybe MMDetection's repository and be named like `tools/mpi_train.sh`, not in `mmcv`'s repository.
> 
> In this pull request, I'm submitting code changes related to `mmcv`. I would be very grateful if you could communicate with your whole team about such a `tools/mpi_train.sh` launching script, as it would be shared by various MM toolboxes just like `tools/slurm_train.sh`. Please tell me where I should contribute this script.

`MASTER_ADDR` is also needed in here for completeness, but openmpi doesn't seem to have environment variable resembles to `SLURM_NODELIST` in `slurm`, should add extra param here for `MASTER_ADDR` or at least add a warning to inform users to set this environment variable by themselves.

Reply:
Should an exception be thrown when `MASTER_ADDR` is not set?

Reply:
> Should an exception be thrown when MASTER_ADDR is not set?

@bryandeng, sorry for late reply, I think it is appropriate.